GameInfo Added

god1385 opened the pull request:
Added GameInfo script, which track the timer and sparkles collected. Added buttons t o start and stop timer. Added Event when player collects Sparkles. If any update needed, feel free to comment

god1385 committed:
GameInfo Added

Added GameInfo script, which track the timer and sparkles collected. Added buttons t o start and stop timer. Added Event when player collects Sparkles. If any update needed, feel free to comment

god1385 committed:
Updated GameInfo

GameInfo now tracks the current level and tracks if the endgame is here. Every time certain number of sparkles are collected, the level is completed and counter goes up

JohnnyZaev commented on Assets/Scripts/GameInfo/GameInfo.cs:
```diff
@@ -0,0 +1,71 @@
+    [SerializeField] private float timerStart;
+    [SerializeField] private int maxSparkles;
+    [SerializeField] private float goodEndingSparkles;
+    [SerializeField] private UnityEvent allSparklesCollected;
```
Нейминг ивентов желательно начинать как и у булевых переменные с предлога. То есть что будет происходить как стартанет ивент. Золотое правило как минимум добавить On -> onSparklesCollected и т.д

JohnnyZaev commented on Assets/Scripts/GameInfo/GameInfo.cs:
```diff
@@ -0,0 +1,71 @@
+    [SerializeField] private int maxSparkles;
+    [SerializeField] private float goodEndingSparkles;
+    [SerializeField] private UnityEvent allSparklesCollected;
+    [SerializeField] private UnityEvent endGame;
```
Same ^

JohnnyZaev commented on Assets/Scripts/GameInfo/GameInfo.cs:
```diff
@@ -0,0 +1,71 @@
+    [SerializeField] private float goodEndingSparkles;
+    [SerializeField] private UnityEvent allSparklesCollected;
+    [SerializeField] private UnityEvent endGame;
+    [SerializeField] private int _maxLevels;
```
Тут поехал нейминг

JohnnyZaev commented on Assets/Scripts/GameInfo/GameInfo.cs:
```diff
@@ -0,0 +1,71 @@
+    private int _levelsCompleted = 0;
+
+
+    private TextMeshProUGUI TimerTextField { get { return timerTextField; } }
```
Не очень понял зачем нужна эта строчка еще и приватная. Есть же такое-же поле сверху. параметры по умолчанию берут модификатор доступа самого поля. То есть у тебя тут гетер только для твоего скрипта.

JohnnyZaev commented on Assets/Scripts/GameInfo/GameInfo.cs:
```diff
@@ -0,0 +1,71 @@
+            TimerTextField.text = Mathf.Round(timerStart).ToString();
+        }
+
+        if (SparklesCounter == maxSparkles)
```
Вот это можно проверять в методе SparklesCounter, чтобы не засорять апдейт. + я тут вижу, что проверяется количество спарклов в комнате. Соответственно я думаю, что это должна проверять каждая комната, а не гейминфо будет засорятся максимумом спарклов на каждую комнату. Тут нужно проверять только то, что все спарклы собраны и именно через процент, а не чтобы прямо идеально все

JohnnyZaev commented on Assets/Scripts/GameInfo/GameInfo.cs:
```diff
@@ -0,0 +1,71 @@
+        if (_levelsCompleted == _maxLevels)
+        {
+            endGame.Invoke();
+        }
```
Тоже самое. Мастерские отправляют инфу собраны они или нет. Там геймменеджер проверяет эту информацию и инвокает ивент. Делать это в апдейте смысла нет

JohnnyZaev commented on Assets/Scripts/WorkShops/WorkShops.cs:
```diff
@@ -16,6 +17,7 @@ void Start()
         }
         _activeWorkshop = _workshops[0];
         _activeWorkshop.SetActive(true);
+        List<int> list = new List<int>();
```
??????

JohnnyZaev reviewed (changes requested):
Фикси по комментам)

god1385 committed:
Updated GameInfo and Added Workshop

GamInfo was updated with naming fixes and removal of levelCompletion check in the Update function. Now WorkShop file is attached to all workshops, where each of them checks  the collection of Sparkles. When certain amount was added, level is considered completed and UnityEvent is send to the GameInfo where LevelCompletedFunction tracks the progress.

JohnnyZaev commented on Assets/Scripts/GameInfo/GameInfo.cs:
```diff
@@ -0,0 +1,65 @@
+public class GameInfo : MonoBehaviour
+{
+    // Added new variable, due to inability of properties to show up in Editor, so property is linked to the variable from the editor
+    public int _levelsCompleted = 0;
```
Надо ли нам видеть эту переменную в инспекторе? Думаю тут нужен атрибут HideInInspector

JohnnyZaev replied:
Ну и публичная переменная с нижним подчеркиванием

JohnnyZaev commented on Assets/Scripts/WorkShops/WorkShops.cs:
```diff
@@ -4,33 +4,32 @@
     [SerializeField] private UnityEvent onWorkshopChanged;
-    private GameObject[] _workshops;
-    private GameObject _activeWorkshop;
+    [SerializeField] private WorkShop[] _workshops;
```
у сериалайз переменных нет нижнего подчеркивания

JohnnyZaev reviewed (changes requested):
Ну совсем мелочи остались)

god1385 committed:
UpdatedNaming

JohnnyZaev commented on Assets/Scripts/GameInfo/GameInfo.cs:
```diff
@@ -0,0 +1,66 @@
+
+public class GameInfo : MonoBehaviour
+{
+    // Added new variable, due to inability of properties to show up in Editor, so property is linked to the variable from the editor
+    [HideInInspector]
+    public int levelsCompleted = 0;
+    [SerializeField] private TextMeshProUGUI timerTextField;
+    [SerializeField] private float timerStart;
+    [SerializeField] private float goodEndingSparkles;
+    [SerializeField] private UnityEvent onEndGame;
+    [SerializeField] private int maxLevels;
```
Почему [HideInInspector] и [SerializeField] стоят в разных местах у переменных?

JohnnyZaev commented on Assets/Scripts/GameInfo/GameInfo.cs:
```diff
@@ -0,0 +1,66 @@
+        }
+    }
+
+    // Update is called once per frame
```
Удали комментарий у апдейта

JohnnyZaev commented on Assets/Scripts/WorkShops/WorkShops.cs:
```diff
@@ -4,33 +4,32 @@
     [SerializeField] private UnityEvent onWorkshopChanged;
-    private GameObject[] _workshops;
-    private GameObject _activeWorkshop;
+    [SerializeField] private WorkShop[] workshops;
```
Почему массив мастерских сериалайз филд?

JohnnyZaev reviewed (changes requested):
Ты обязательно попадешь в меин)

god1385 committed:
Updated GameInfo and Workshops

JohnnyZaev reviewed (approved):
гц
